Added clearer error message for tracers in numpy.split

gnecula opened the pull request:
Fixes: #2506

Now we print:

TypeError: In jax.numpy.split argument 2: tracer value encountered where a
concrete value is expected.
Use `functools.partial` or transformation parameters such as `static_argnums`
for `jit` to preserve the concrete value during JAX tracing.
See `https://jax.readthedocs.io/en/latest/jaxpr.html#tracer_must_be_concrete`.
Encountered value: Traced<ShapedArray(float32[],
weak_type=True):JaxprTrace(level=-1/1)>

gnecula committed:
Added clearer error message for tracers in numpy.split

Now we print:

ConcretizationTypeError: Abstract tracer value where concrete value is expected (in
jax.numpy.split argument 1).
Use transformation parameters such as `static_argnums` for `jit` to avoid
tracing input values.
See `https://jax.readthedocs.io/en/latest/faq.html#abstract-tracer-value-where-concrete-value-is-expected-error`.
Encountered value: Traced<ShapedArray>

mattjj commented on jax/core.py:
```diff
@@ -653,6 +653,21 @@ def error(self, *args):
+         "Encountered value: {}")
+  if isinstance(arg, Tracer):
+    raise TypeError(msg.format(context, arg))
+  elif isinstance(arg, Sequence):
```
Do we need this polymorphism? Maybe we shouldn't handle sequences here, and leave that to the caller.

gnecula replied:
Good point, I can add this feature to the call site. I suspect we'll use this function in more places, we can see later if it needs more functionality.

mattjj reviewed (changes requested):
Thanks for working on this!

The FAQ text and the function `check_concrete` are conflating when things are tracers versus when they are concrete in the abstract value lattice. The two are independent in that something can be a Tracer instance yet totally concrete, as we rely on for autodiff. We should probably keep them straight.

For example, if one tweaks `make_jaxpr` not to raise to the shaped level (by replacing `xla.abstractify` with `core.get_aval`), and also disables an exception I inserted for remat (which doesn't actually signal erroneous behavior), this works:

```python
In [1]: from jax import make_jaxpr

In [2]: import jax.numpy as np

In [3]: make_jaxpr(lambda i: print(i) or np.split(np.zeros(3), i))(3)
Traced<ConcreteArray(3, weak_type=True):JaxprTrace(level=0/0)>
Out[3]:
{ lambda b c d ; a.
  let
  in (b, c, d) }
```

For that reason, I recommend:
1. renaming `check_concrete` to `check_tracer` (and since I also recommend removing the sequence polymorphism, it'll just be `check_tracer = lambda x: isinstance(x, Tracer)`, so I'm not sure we need it at all)
2. changing the FAQ language not to suggest tracers are the issue when it's actually about abstraction level
3. change the `numpy.split` check to be one about abstraction / specialization level rather than checking for tracer instances

gnecula commented:
> The FAQ text and the function `check_concrete` are conflating when things are tracers versus when they are concrete in the abstract value lattice. The two are independent in that something can be a Tracer instance yet totally concrete, as we rely on for autodiff. We should probably keep them straight.

I've been struggling to find the right vocabulary to talk about these concepts (and I used "concrete" in a different way that you are using it). Let's try to find a vocabulary that is also relatively easy to explain to users. How about the following: 

A **tracer** value is introduced for the argument of a JAX-transformed function, except the arguments identified by special parameters such as `static_argnums` for `jit` or `static_broadcasted_argnums` for `pmap`. Typically, computations that involve at least a tracer value will produce a tracer value. Besides tracer values, there are **regular** Python values: values that are computed outside JAX transformations, or arise from above-mentioned static arguments of certain JAX transformations, or computed solely from other regular Python values. 

A tracer value carries an **abstract value**, e.g., ShapedArray with information about the shape and dtype of an array. Some tracers, e.g., those that are introduced for arguments of autodiff transformations, carry ConcreteArray abstract values that actually include the actual array data. The same is true for tracer values computed from these, perhaps in combination with regular values. A **concrete** value is either a regular value or a tracer value with ConcreteArray abstract value. 

Most often values computed from tracer values are themselves tracer values. There are very few exceptions, when a computation can be entirely done using the abstract value carried by a tracer, in which case the result can be a regular value. For example, getting the shape of a tracer with ShapedArray abstract value. Another example, is when explicitly casting a concrete tracer value to a regular type: `int(x)`. 

WDYT? 

I personally would prefer if the users need not understand the "concrete" aspect, and worry only about tracer vs. regular value and how to control that. The "concrete" aspect is difficult to explain and to understand and is somewhat non-compositional: in `lambda x: x + 1` the value of `x` will be a concrete tracer if we use `grad` on that function, unless somewhere in a higher scope we use `jit` on the enclosing function.  

I also would like a better description of when you can get a regular value from a tracer: e.g., from a concrete tracer sometimes you get a concrete tracer and sometimes you get a regular value. Is it only "shape" and explicit casting that gets a regular value?

gnecula commented:
> The FAQ text and the function `check_concrete` are conflating when things are tracers versus when they are concrete in the abstract value lattice. The two are independent in that something can be a Tracer instance yet totally concrete, as we rely on for autodiff. We should probably keep them straight.

My goal here is to have a strategy for systematically reporting API usage errors, making them clear and actionable. How about this: 

If a numpy function is called with wrong regular values, we must report the same error as numpy. This can often be achieved easily by actually calling the original numpy function. If the numpy function is called with concrete tracer values, the behavior should be the same, including error messages, as when called with the carried regular values. If the numpy function is called with non-concrete tracer values in unexpected (or unimplemented) parameter positions, then there should be an error message explaining that the particular parameter of the particular function must be concrete, and a link to FAQ where we explain that one can use partially evaluated functions or `static_argnums` for `jit`.  

WDYT?

gnecula commented:
> For example, if one tweaks `make_jaxpr` not to raise to the shaped level (by replacing `xla.abstractify` with `core.get_aval`), and also disables an exception I inserted for remat (which doesn't actually signal erroneous behavior), this works:
> 
> ```python
> In [1]: from jax import make_jaxpr
> 
> In [2]: import jax.numpy as np
> 
> In [3]: make_jaxpr(lambda i: print(i) or np.split(np.zeros(3), i))(3)
> Traced<ConcreteArray(3, weak_type=True):JaxprTrace(level=0/0)>

A more realistic scenario is to replace make_jaxpr with grad: the index is a concrete tracer. If we use `np.split(arr, i)` then numpy internally properly converts `i` to an integer and the code works. However, if we use `np.split(arr, [1, i])` then we get an error in numpy `TypeError: slice indices must be integers or None or have an __index__ method`. The point is that we cannot always rely on numpy to be systematic about converting to integer. Maybe this is an omission in numpy? 

I do think we need to add error checking for some of the numpy functions. I have changed the code to check for concreteness instead of tracers. 

gnecula commented:
> For that reason, I recommend:
> 
> 1. renaming `check_concrete` to `check_tracer` (and since I also recommend removing the sequence polymorphism, it'll just be `check_tracer = lambda x: isinstance(x, Tracer)`, so I'm not sure we need it at all)

I do think that we need a function, to construct the proper error message. 

> 2. changing the FAQ language not to suggest tracers are the issue when it's actually about abstraction level

I will change the FAQ once we agree on the language. 

> 3. change the `numpy.split` check to be one about abstraction / specialization level rather than checking for tracer instances

Done

gnecula commented:
@mattjj PTAL

mattjj commented:
(Sorry it took me 9 days to respond :/)

IMO everything up until "WDYT?" in [your first comment](https://github.com/google/jax/pull/2508#issuecomment-604360901) is excellent and very well put.

I'd perhaps add, as another example, that in addition to `int(x)` producing a Python int when concreteness makes it possible, `bool(x)` will produce a Python bool when concreteness makes it possible. That case is especially salient because of how often it arises in control flow.

> I personally would prefer if the users need not understand the "concrete" aspect, and worry only about tracer vs. regular value and how to control that.

This part I disagree with though, because it's the abstraction level and not tracer-ness that makes this fail:

```python
def foo(x):
  return x if x > 0 else 0.

jit(foo)(3.)
```

while this works:

```python
grad(foo)(3.)
```

As for the case of `jnp.split`, I've revised my thinking based on this discussion; see below.

> I also would like a better description of when you can get a regular value from a tracer: e.g., from a concrete tracer sometimes you get a concrete tracer and sometimes you get a regular value. Is it only "shape" and explicit casting that gets a regular value?

I agree it would be good to specify this precisely. I think it's exactly that we get tracers out of binding primitives, i.e. applying `Primitive.bind` to Tracer arguments. When `x` is a Tracer instance, `x.shape` doesn't bind a primitive, just like `bool(x)` or `len(x)` or `int(x)` don't bind a primitive. Those all return regular Python values (a tuple, boolean, integer, and integer, respectively).

Ultimately this is a special list of operations that we construe as Python builtins operating on abstract value information during abstract interpretation, as opposed to operations in the embedded/object language (corresponding to operations on Tracers). It's a whitelist in the sense that `onp.array(x, dtype=onp.int32)` is always an error if `x` is a Tracer, even though it's analogous to the Python builtin `int`. We just don't define that operation to be part of our abstract interpretation of Python.

In discussing with @dougalm, we came up with three motivations for functions/methods on Tracers producing non-Tracer values (i.e. for considering an operation to be part of abstract interpretation):
1. when we query shape/dtype information (always part of the abstract evaluation, not object language operations)
2. when we want a concrete value for Python control, which is why we do it for `bool` and `int` (the builtin `range` used to call `int` on its argument, but doesn't in Python 3 unfortunately)
3. when we promote values to shape information, like `jnp.split(x, n)` does to `n`

We think (1) is 10x more important than (2), and (2) is 10x more important than (3). Actually, we think we can probably remove (3) from the system if we drop out of traces (i.e. call core.full_lower) more eagerly (there are a few missed opportunities). Like, when should we have a Traced concrete integer anyway? (Maybe just for `jax.remat`...) And if we did that, then we _could_ make this `jnp.split` error about Tracers!

I think we have to keep (2). If the alternative is to have just Tracers and Python values, we can't explain why `grad(lambda x: x if x > 0 else 0.)` works.

(This conversation (including with @dougalm) has made me realize that we are probably doing the _wrong_thing to do this for other builtins like `float` and `complex`, because `grad(lambda x: float(x) * 2.)(2.)` produces a zero right now.)

So, bottom line: I think we still need to distinguish abstract value errors from Tracer errors, with `grad(lambda x: x if x > 0 else 0.)` being the sharp example. But as for `jnp.split(x, n)` for a Tracer `n` I think we _could_ raise a Tracer error if it's more useful / understandable to the user (even if we perform the check using `int(n)` for convenience).

As @dougalm put it for the `jnp.split` case:
> the overwhelmingly likely case is that the user was tracing the value with a non-concrete aval
> and the fact that if they were tracing it with a concrete aval then it would be ok seems like more detail than needed

WDYT?

> The point is that we cannot always rely on numpy to be systematic about converting to integer. Maybe this is an omission in numpy? 

I think to be consistent we should treat `__index__` like we do `__int__`.

mattjj commented on docs/faq.rst:
```diff
@@ -50,27 +49,76 @@ Without `jit` the output is::
+.. comment We refer to the anchor below in JAX error messages
+.. _value_must_be concrete:
+
+`Tracer value where concrete value is expected` error
```
I still don't think we should set up "Tracer" and "concrete" as mutually exclusive alternatives. How about instead "Tracer where Python value is expected" as one error, and "Abstract value where concrete value is expected" as another, or similar? The former can be used for `jnp.split(x, n)` for a Tracer `n`, while the latter can be used for `jit(lambda x: x if x > 0 else 0.)(3.)` (and won't be raised on `grad(lambda x: x if x > 0 else 0.)(3.)` even though in both cases `x` refers to a Tracer instance in the body of the function being transformed).

mattjj commented on docs/faq.rst:
```diff
@@ -50,27 +49,76 @@ Without `jit` the output is::
+    See `https://jax.readthedocs.io/en/latest/faq.html#value_must_be_concrete`.
+    Encountered value: Traced<ShapedArray(int32[], weak_type=True):JaxprTrace(level=-1/1)>
+
+This would happen with other JAX transformations, e.g., :func:`jax.grad`.
```
This sentence isn't true though; these work with `grad`. How about deleting this sentence?

mattjj commented on docs/faq.rst:
```diff
@@ -50,27 +49,76 @@ Without `jit` the output is::
+
+This would happen with other JAX transformations, e.g., :func:`jax.grad`.
+We must change the way we use :func:`jax.jit` to ensure that the ``num_sections`` and ``axis`` arguments use
+their concrete values (``2`` and ``0`` respectively). For example, we can apply the transformation to a closure
```
If we keep "concrete" to be about abstraction level and not to mean "not a Tracer", which I think we must, then we should revise this sentence to be about keeping ``2`` and ``0`` as non-Tracer values.

mattjj commented on jax/core.py:
```diff
@@ -639,21 +639,38 @@ def __repr__(self): return '*'
+class ConcretizationTypeError(TypeError): pass
+
+def raise_concretization_error(val, context=""):
+  msg = (f"Abstract tracer value encountered where a concrete value is expected ({context}).\n"
```
How about: "Tracer value (at an abstraction level higher than concrete) encountered where a concrete value is expected."?

mattjj commented on jax/core.py:
```diff
@@ -639,21 +639,38 @@ def __repr__(self): return '*'
+def raise_concretization_error(val, context=""):
+  msg = (f"Abstract tracer value encountered where a concrete value is expected ({context}).\n"
+          "Use `functools.partial` or transformation parameters such as `static_argnums` for `jit` "
+          "to preserve the concrete value during JAX tracing.\n"
```
Instead of "to preserve the concrete value" how about "to avoid tracing some input values"?

mattjj commented on docs/faq.rst:
```diff
@@ -50,27 +49,76 @@ Without `jit` the output is::
+their concrete values (``2`` and ``0`` respectively). For example, we can apply the transformation to a closure
+that encapsulates the arguments to be protected, either manually as below or by using ``functools.partial``::
+
+  jax.jit(lambda arr: split(arr, 2, 0))(np.zeros(4))
```
This isn't a great pattern to encourage because it'll create a fresh function argument to `jit` on every invocation and thus prevent any compilation caching. Using `static_argnums` is better for caching, and is probably how we should recommend ensuring that some inputs to a `jit`ted function aren't traced.

mattjj commented on jax/core.py:
```diff
@@ -639,21 +639,38 @@ def __repr__(self): return '*'
+
+def raise_concretization_error(val, context=""):
+  msg = (f"Abstract tracer value encountered where a concrete value is expected ({context}).\n"
+          "Use `functools.partial` or transformation parameters such as `static_argnums` for `jit` "
```
Using `functools.partial` and lexical closure are less good for `jit` because they can lead to more recompilation than `static_argnums` (by defining new function objects every time). How about we just recommend `static_argnums` here?

mattjj commented:
Actually, @dougalm pointed out to me that if we did a good job of `core.full_lower`-ing, not only could we remove motivation (3) but also motivation (2). In that case, the example `grad(lambda x: x if x > 0 else 0.)` _could_ be about a Tracer getting through (because `x > 0` will be a Tracer iff its value is not concrete, since we'd drop out of the JVPTracer).

However, I wouldn't want that to inform how we report errors until we actually make that change to the system. It's possibly details like `jax.remat` will make it not work. So my recommendations are the same, though there's more thread to pull here, and if we really wanted to only ever raise Tracer errors, I think we know what we have to try.

gnecula commented:
I have finally gotten back to this. I have done the following:

  * error message is "Abstract tracer encountered where concrete value is expected".
  * I have added `core.concrete_or_error`(int, x, "argument 1 of split")`. This is needed
    when `onp` does not use `int()` inside but gives out random errors, such as for `split`
 * I have written two FAQ entries, one specifically for this error and one called "Different kinds of JAX values"

gnecula committed:
Fixed tests, slight change to the error message

gnecula committed:
Expanded the FAQ entry about abstract tracers for higher-order primitives

jekbradbury commented on docs/faq.rst:
```diff
@@ -65,6 +65,133 @@ Additional reading:
+    except those denoted by ``static_broadcasted_argnums``.
+  * :func:`jax.vmap`, :func:`jax.make_jaxpr`, :func:`xla_computation`:
+    introduce **abstract tracers** for all positional arguments.
+  * :func:`jax.jvp` and :func:`jax.grad` introduce **concrete tracers**
```
This isn't quite true, because `grad` might be called inside a `jit`, in which case the trace happens at shape level. It's more accurate to talk about levels of abstraction and say that transformations like `jit` and `vmap` raise the abstraction level to "shaped" while `jvp` doesn't raise it.

gnecula replied:
I have added additional text to the explanation for `grad`.

mattjj replied:
I agree with James that it's better to say "raise the level of abstraction". For example, instead of "introduces abstract tracers" how about "introduces tracers with abstract values raised to at least the ____ level", where ____ is different for different transformations?

jekbradbury commented:
I guess I don't really buy that making the error here about tracers rather than abstraction makes it easier to understand. If our tracing machinery makes user code fail for reasons other than excessive abstraction or side effects that seems like a bug?

In general I'm not sure that the FAQ entry about "different kinds of JAX values" helps users build the right mental model; as much as possible, "mathy" code shouldn't care whether a value is traced but rather whether the abstraction level is too high. In parts of the documentation where we do talk about tracers (e.g. explaining their behavior under `print`) I'd prefer something that makes their role more explicit: they flow through user code while exposing an interface (their aval) that stands in for sets of Python values ("abstract interpretation"), and they carry metadata that allows adding to or changing the meaning of the code ("alternative interpretation").

gnecula commented:
> I guess I don't really buy that making the error here about tracers rather than abstraction makes it easier to understand. If our tracing machinery makes user code fail for reasons other than excessive abstraction or side effects that seems like a bug?

The error is about "abstract tracers", so I think we are in agreement. Are you objecting about using the word "tracer"? But the way the user can fix this is most likely to avoid tracers.

gnecula commented:
> In general I'm not sure that the FAQ entry about "different kinds of JAX values" helps users build the right mental model; as much as possible, "mathy" code shouldn't care whether a value is traced but rather whether the abstraction level is too high.

I do think that sooner or later users will end up with tracers in non mathy core, and it is worth explaining what tracers are, and how to control them.

gnecula committed:
Added clarification for tracers inside jit of grad

mattjj commented on docs/faq.rst:
```diff
@@ -65,6 +65,136 @@ Additional reading:
+using the abstract value carried by a tracer, in which case the result
+can be a regular value. For example, getting the shape of a tracer
+with ``ShapedArray`` abstract value. Another example, is when explicitly
+casting a concrete tracer value to a regular type: ``int(x)`` or ``float(x)``.
```
We should probably remove `float(x)` as an example, or decide how to change its behavior (make it equivalent to `x.astype('float')`?) as per the discussion in #2783 .

gnecula replied:
Right, I learned this after I wrote the note. I replaced with x.astype.

mattjj commented on docs/faq.rst:
```diff
@@ -65,6 +65,136 @@ Additional reading:
+
+Here is how the transformations introduce abstract or concrete tracers:
+
+  * :func:`jax.jit`: introduces **abstract tracers** for all positional arguments
```
Instead of "abstract tracers" how about "tracers with abstract values at the Shaped level"?

mattjj replied:
Same below.

gnecula replied:
Added definition above

mattjj commented on docs/faq.rst:
```diff
@@ -65,6 +65,136 @@ Additional reading:
+concreteness makes it possible. That case is especially salient because
+of how often it arises in control flow.
+
+Here is how the transformations introduce abstract or concrete tracers:
```
I think "abstract tracers" really means "tracers with abstract values raised above the concrete level." Maybe it'd be good to define that term precisely?

gnecula replied:
The text above said:

> A tracer value carries an **abstract** value, e.g., ``ShapedArray`` with information
> about the shape and dtype of an array.

I have added 
> We will refer here to such tracers as **abstract tracers**.

So that I can refer to them as "abstract tracers". I have not used the word "raised" because 
I do not think it is necessary here to talk about abstraction levels and why a level is higher than others.

mattjj reviewed (approved):
Overall looks great to me. I have some suggested tweaks to the language, but if you want to land and iterate we can do that too!

gnecula committed:
Updated FAQ language in response to reviews

fasghq commented:
Maybe I miss something obvious, but what should I do if I need to use inequation inside the function for `jax.jit`? Here's my function:
`def tt_round_model(tt, delta):`
`...`
`for x in s:`
`if x < delta:`
`...`
As I found, I should add `static_argnums` in such case, so I added `delta` to it:
`fast_tt_round = jit(tt_round_model, static_argnums=(1))`
`ttr = fast_tt_round(tts_s, eps / math.sqrt(len(tts_s.core) - 1))`
But I still have the problem in compilation:
![Screenshot at 23-41-52](https://user-images.githubusercontent.com/62699960/87351676-747c6e00-c562-11ea-8142-e3c69885bfda.png)
I suppose that the problem is in my `x` variable which is diagonal element of `s` matrix which I get inside this algorithm, so I can't add it to `static_argnums`. 
Thank you for your help!

jakevdp commented:
A merged PR is not a great place to bring up a usage question. Feel free to open a new issue to get assistance!
